File: Utils/save_map_point/src/egvg_vis.cpp
#include <ros/ros.h>
#include <nav_msgs/OccupancyGrid.h>
#include <sensor_msgs/PointCloud2.h>
#include <visualization_msgs/MarkerArray.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl/io/pcd_io.h>
#include <pcl/filters/voxel_grid.h>
#include <pcl_conversions/pcl_conversions.h>

class OccupancyGridToPointCloud {
public:
    OccupancyGridToPointCloud(ros::NodeHandle& nh) : nh_(nh) {
        // 加载全局地图
        std::string pcd_file;
        nh_.param<std::string>("pcd_file", pcd_file, "global_map.pcd");
        nh_.param<double>("offset_x", offset_x_, 0.0);
        nh_.param<double>("offset_y", offset_y_, 0.0);
        if (pcl::io::loadPCDFile<pcl::PointXYZ>(pcd_file, *global_map_)) {
            ROS_ERROR("Can not loca PCD file : %s", pcd_file.c_str());
            return;
        }
        ROS_INFO("Loaded PCD file: %s", pcd_file.c_str());

        // 对全局地图进行降采样
        double voxel_size;
        nh_.param<double>("voxel_size", voxel_size, 0.1);
        pcl::VoxelGrid<pcl::PointXYZ> voxel_filter;
        voxel_filter.setInputCloud(global_map_);
        voxel_filter.setLeafSize(voxel_size, voxel_size, voxel_size);
        voxel_filter.filter(*downsampled_map_);
        // ROS_INFO("全局地图降采样完成，点数: %lu", downsampled_map_->size());

        // 发布全局地图点云
        global_map_pub_ = nh_.advertise<sensor_msgs::PointCloud2>("/egvg_vis/global_map", 1);
        // 发布点云
        free_cloud_pub_ = nh_.advertise<sensor_msgs::PointCloud2>("/egvg_vis/free_cloud", 1);
        occupy_cloud_pub_ = nh_.advertise<sensor_msgs::PointCloud2>("/egvg_vis/occupy_cloud", 1);

        // 发布修改后的 MarkerArray
        modified_marker_pub_ = nh_.advertise<visualization_msgs::MarkerArray>("/egvg_vis/voronoi_markers", 1);

        // 发布 voronoi 点云
        voronoi_cloud_pub_ = nh_.advertise<sensor_msgs::PointCloud2>("/egvg_vis/voronoi_cloud", 1);

        // 订阅 /
        occupancy_grid_sub_ = nh_.subscribe("/sdf_map/occupancy_2D", 1, &OccupancyGridToPointCloud::occupancyGridCallback, this);
        // 订阅 /voronoi/gridmap
        voronoi_grid_sub_ = nh_.subscribe("/voronoi/gridmap", 1, &OccupancyGridToPointCloud::voronoiGridCallback, this);
        // 订阅 /voronoi/gvg_markers
        voronoi_marker_sub_ = nh_.subscribe("/voronoi/gvg_markers", 1, &OccupancyGridToPointCloud::voronoiMarkerCallback, this);


        // 定时器，用于以 1Hz 发布全局地图
        timer_ = nh_.createTimer(ros::Duration(1.0), &OccupancyGridToPointCloud::publishGlobalMap, this);
    }

private:
    ros::NodeHandle nh_;
    ros::Publisher global_map_pub_;
    ros::Publisher free_cloud_pub_;
    ros::Publisher occupy_cloud_pub_;
    ros::Publisher modified_marker_pub_;
    ros::Publisher voronoi_cloud_pub_;
    ros::Subscriber voronoi_grid_sub_;
    ros::Subscriber voronoi_marker_sub_;
    ros::Subscriber occupancy_grid_sub_;
    ros::Timer timer_;

    pcl::PointCloud<pcl::PointXYZ>::Ptr global_map_{new pcl::PointCloud<pcl::PointXYZ>()};
    pcl::PointCloud<pcl::PointXYZ>::Ptr downsampled_map_{new pcl::PointCloud<pcl::PointXYZ>()};

    double offset_x_, offset_y_;

    void publishGlobalMap(const ros::TimerEvent&) {
        sensor_msgs::PointCloud2 cloud_msg;
        pcl::toROSMsg(*downsampled_map_, cloud_msg);
        cloud_msg.header.frame_id = "world";
        cloud_msg.header.stamp = ros::Time::now();
        global_map_pub_.publish(cloud_msg);
    }

    void occupancyGridCallback(const nav_msgs::OccupancyGrid::ConstPtr& msg) {
        pcl::PointCloud<pcl::PointXYZ>::Ptr free_cloud(new pcl::PointCloud<pcl::PointXYZ>());
        pcl::PointCloud<pcl::PointXYZ>::Ptr occupy_cloud(new pcl::PointCloud<pcl::PointXYZ>());

        double resolution = msg->info.resolution;
        double origin_x = msg->info.origin.position.x;
        double origin_y = msg->info.origin.position.y;

        int width = msg->info.width;
        int height = msg->info.height;

        // 遍历 OccupancyGrid 数据
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                int index = y * width + x;
                double world_x = origin_x + (x + 0.5) * resolution + offset_x_;
                double world_y = origin_y + (y + 0.5) * resolution + offset_y_;

                if (msg->data[index] == 0) {  // free 点
                    for (double z = 0.0; z <= 0.0; z += 0.2) {
                        free_cloud->points.emplace_back(world_x, world_y, z);
                    }
                }
                if (msg->data[index] == -1)
                {
                    for (double z = 0.0; z < 2.0; z += 0.2)
                    {
                        occupy_cloud->points.emplace_back(world_x, world_y, z);
                    }
                }
            }
        }

        // 发布 点云
        sensor_msgs::PointCloud2 free_cloud_msg;
        pcl::toROSMsg(*free_cloud, free_cloud_msg);
        free_cloud_msg.header.frame_id = "world";
        free_cloud_msg.header.stamp = ros::Time::now();
        free_cloud_pub_.publish(free_cloud_msg);

        sensor_msgs::PointCloud2 occupy_cloud_msg;
        pcl::toROSMsg(*occupy_cloud, occupy_cloud_msg);
        occupy_cloud_msg.header.frame_id = "world";
        occupy_cloud_msg.header.stamp = ros::Time::now();
        occupy_cloud_pub_.publish(occupy_cloud_msg);
    }


    void voronoiGridCallback(const nav_msgs::OccupancyGrid::ConstPtr& msg) {
        pcl::PointCloud<pcl::PointXYZ>::Ptr voronoi_cloud(new pcl::PointCloud<pcl::PointXYZ>());

        double resolution = msg->info.resolution;
        double origin_x = msg->info.origin.position.x;
        double origin_y = msg->info.origin.position.y;

        int width = msg->info.width;
        int height = msg->info.height;

        // 遍历 OccupancyGrid 数据
        for (int y = 0; y < height; ++y) {
            for (int x = 0; x < width; ++x) {
                int index = y * width + x;
                double world_x = origin_x + (x + 0.5) * resolution;
                double world_y = origin_y + (y + 0.5) * resolution;

                if (msg->data[index] == -128) {  // voronoi 点
                    voronoi_cloud->points.emplace_back(world_x, world_y, 0.0);
                } 
            }
        }
        
        // 发布 voronoi 点云
        sensor_msgs::PointCloud2 voronoi_cloud_msg;
        pcl::toROSMsg(*voronoi_cloud, voronoi_cloud_msg);
        voronoi_cloud_msg.header.frame_id = "world";
        voronoi_cloud_msg.header.stamp = ros::Time::now();
        voronoi_cloud_pub_.publish(voronoi_cloud_msg);
    }

    void voronoiMarkerCallback(const visualization_msgs::MarkerArray::ConstPtr& msg) {
        visualization_msgs::MarkerArray modified_markers;

        for (const auto& marker : msg->markers) {
            visualization_msgs::Marker modified_marker = marker;

            if (marker.color.g == 1.0 && marker.color.r == 0.0 && marker.color.b == 0.0) {
                // 修改颜色为红色
                modified_marker.color.r = 1.0;
                modified_marker.color.g = 0.0;
                modified_marker.color.b = 0.0;
            } else if (marker.color.r == 0.0 && marker.color.g == 0.0 && marker.color.b == 1.0) {
                // 修改颜色为绿色
                modified_marker.color.r = 0.0;
                modified_marker.color.g = 1.0;
                modified_marker.color.b = 0.0;
            }

            // 修改大小
            modified_marker.scale.x = 0.6;
            modified_marker.scale.y = 0.6;
            modified_marker.scale.z = 0.6;

            modified_markers.markers.push_back(modified_marker);
        }

        // 发布修改后的 MarkerArray
        modified_marker_pub_.publish(modified_markers);
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "egvg_vis");
    ros::NodeHandle nh("~");

    OccupancyGridToPointCloud node(nh);

    ros::spin();
    return 0;
}

File: Utils/save_map_point/src/egvg_vis.py
#!/usr/bin/env python3
import rospy
import numpy as np
from nav_msgs.msg import OccupancyGrid
from sensor_msgs.msg import PointCloud2, PointField
from visualization_msgs.msg import MarkerArray, Marker
import sensor_msgs.point_cloud2 as pc2

class OccupancyGridToPointCloud:
    def __init__(self):
        # 初始化节点
        rospy.init_node("occupancy_grid_to_pointcloud", anonymous=True)

        # 从参数服务器获取分辨率
        self.resolution = rospy.get_param("~resolution", 0.1)  # 默认分辨率为 0.1
        self.z_max = rospy.get_param("~z_max", 2.0)  # 最大高度
        self.z_min = rospy.get_param("~z_min", 0.0)  # 最小高度

        # 订阅 OccupancyGrid 消息
        self.occupancy_grid_sub = rospy.Subscriber("/sdf_map/occupancy_2D", OccupancyGrid, self.occupancy_grid_callback)
        self.voronoi_grid_sub = rospy.Subscriber("/voronoi/gridmap", OccupancyGrid, self.voronoi_grid_sub_callback)
        self.voronoi_marker_sub = rospy.Subscriber("/voronoi/gvg_markers", MarkerArray, self.voronoi_marker_callback)

        # 发布 PointCloud2 消息
        self.obs_pointcloud_pub = rospy.Publisher("/voronoi/obs_cloud", PointCloud2, queue_size=10)
        self.voronoi_grid_pub = rospy.Publisher("/voronoi/gridmap_pointcloud", PointCloud2, queue_size=10)
        self.gvg_marker_pub = rospy.Publisher("/voronoi/gvg_markers_vis", MarkerArray, queue_size=10)

    def voronoi_marker_callback(self, msg):
        modified_markers = MarkerArray()

        for marker in msg.markers:
            # 检查是否是绿色点
            if marker.color.g == 1.0 and marker.color.r == 0.0 and marker.color.b == 0.0:
                # 修改颜色为红色
                marker.color.r = 1.0
                marker.color.g = 0.0
                marker.color.b = 0.0

                # 修改大小为 0.6
                marker.scale.x = 0.6
                marker.scale.y = 0.6
                marker.scale.z = 0.6
            
            elif marker.color.r == 0.0 and marker.color.g == 0.0 and marker.color.b == 1.0:
                # 修改颜色为绿色
                marker.color.r = 0.0
                marker.color.g = 1.0
                marker.color.b = 0.0

                # 修改大小为 0.6
                marker.scale.x = 0.6
                marker.scale.y = 0.6
                marker.scale.z = 0.6

            # 将修改后的 Marker 添加到新的 MarkerArray 中
            modified_markers.markers.append(marker)

        # 发布修改后的 MarkerArray
        self.gvg_marker_pub.publish(modified_markers)


    def voronoi_grid_sub_callback(self, msg):
        """
        处理 VoronoiGrid 消息，将 data 值等于 -1 的网格转换为点云
        """
        # 获取地图信息
        resolution = msg.info.resolution
        width = msg.info.width
        height = msg.info.height
        origin_x = msg.info.origin.position.x
        origin_y = msg.info.origin.position.y

        # 初始化点云列表
        points = []

        # 遍历 VoronoiGrid 数据
        for y in range(height):
            for x in range(width):
                index = y * width + x
                if msg.data[index]  == -128:  # 只处理 data 值等于 128 的网格
                    # print(f"map data: {msg.data[index]}")
                    # 计算网格中心的世界坐标
                    world_x = origin_x + (x + 0.5) * resolution
                    world_y = origin_y + (y + 0.5) * resolution

                    z = self.z_min
                    points.append([world_x, world_y, z])
        # 将点云转换为 PointCloud2 消息
        pointcloud_msg = self.create_pointcloud2_msg(points, msg.header)
        # 发布点云
        self.voronoi_grid_pub.publish(pointcloud_msg)


    def occupancy_grid_callback(self, msg):
        """
        处理 OccupancyGrid 消息，将 data 值等于 -1 的网格转换为点云
        """
        # 获取地图信息
        resolution = msg.info.resolution
        width = msg.info.width
        height = msg.info.height
        origin_x = msg.info.origin.position.x
        origin_y = msg.info.origin.position.y

        # 初始化点云列表
        points = []

        # 遍历 OccupancyGrid 数据
        for y in range(height):
            for x in range(width):
                index = y * width + x
                if msg.data[index] == -1:  # 只处理 data 值等于 -1 的网格
                    # 计算网格中心的世界坐标
                    world_x = origin_x + (x + 0.5) * resolution
                    world_y = origin_y + (y + 0.5) * resolution

                    # 生成从 z_min 到 z_max 的多层点
                    z = self.z_min
                    while z <= self.z_max:
                        points.append([world_x, world_y, z])
                        z += self.resolution * 2

        # 将点云转换为 PointCloud2 消息
        pointcloud_msg = self.create_pointcloud2_msg(points, msg.header)

        # 发布点云
        self.obs_pointcloud_pub.publish(pointcloud_msg)

    def create_pointcloud2_msg(self, points, header):
        """
        将点云列表转换为 PointCloud2 消息
        """
        fields = [
            PointField("x", 0, PointField.FLOAT32, 1),
            PointField("y", 4, PointField.FLOAT32, 1),
            PointField("z", 8, PointField.FLOAT32, 1),
        ]
        pointcloud_msg = pc2.create_cloud(header, fields, points)
        return pointcloud_msg

if __name__ == "__main__":
    try:
        node = OccupancyGridToPointCloud()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass